fpgaconvnet/models/modules/Module.py

```diff
@@ -8,275 +8,234 @@
+    name: ClassVar[str]
+    backend: ClassVar[BACKEND]
+    dimensionality: ClassVar[set[DIMENSIONALITY]]
+    register: ClassVar[bool] = False
+    repetitions: int = 1
+
+    @property
+    @abstractmethod
+    def input_ports(self) -> list[Port]:
+        pass
+
+    @property
+    @abstractmethod
+    def output_ports(self) -> list[Port]:
+        pass
+
+    @property
+    @abstractmethod
+    def input_iter_space(self) -> list[list[int]]:
+        pass
+
+    @property
+    @abstractmethod
+    def output_iter_space(self) -> list[list[int]]:
+        pass
+
+    @property
+    @abstractmethod
+    def rate_in(self) -> list[float]:
+        pass
+
+    @property
+    @abstractmethod
+    def rate_out(self) -> list[float]:
+        pass
+
+    @property
+    def ports_in(self) -> int:
+        return len(self.input_ports)
+
+    @property
+    def ports_out(self) -> int:
+        return len(self.output_ports)
+
+    @property
+    def input_simd_lanes(self) -> list[list[int]]:
+        return [ p.simd_lanes for p in self.input_ports ]
+
+    @property
+    def output_simd_lanes(self) -> list[list[int]]:
+        return [ p.simd_lanes for p in self.output_ports ]
+
+    @abstractmethod
+    def functional_model(self, data: np.ndarray) -> np.ndarray:
+        pass
+
+    @abstractmethod
+    def resource_parameters(self) -> list[int]:
+        pass
+
+    @abstractmethod
+    def resource_parameters_heuristics(self) -> dict[str, list[int]]:
+        pass
+
+    @abstractmethod
+    def pipeline_depth(self) -> int:
+        pass
+
+    def get_rate_in(self, idx: int) -> float:
+        assert idx < self.ports_in, "Invalid input port index"
+        return self.rate_in[idx]
+
+    def get_rate_out(self, idx: int) -> float:
+        assert idx < self.ports_out, "Invalid output port index"
+        return self.rate_out[idx]
+
+    def get_input_simd_lanes(self, idx: int) -> list[int]:
+        assert idx < self.ports_in, "Invalid input port index"
+        return self.input_ports[idx].simd_lanes
+
+    def get_output_simd_lanes(self, idx: int) -> list[int]:
+        assert idx < self.ports_out, "Invalid output port index"
+        return self.output_ports[idx].simd_lanes
+
+    def latency(self) -> int:
+        latency_in = max([int(np.prod(self.input_iter_space[i]) \
+                / self.rate_in[i]) for i in range(self.ports_in)])
+        latency_out = max([int(np.prod(self.output_iter_space[i]) \
+                / self.rate_out[i]) for i in range(self.ports_out)])
+        return max(latency_in, latency_out)
+
+    def module_info(self) -> dict:
+        info = asdict(self)
+        info["name"] = self.name
+        info["backend"] = self.backend.name
+        info["dimensionality"] = [ d.value for d in self.dimensionality ]
+        return info
+
+    def resource_parameters_heuristics(self) -> Dict[str, int]:
+        return {}
+
+@dataclass(kw_only=True)
+class ModuleChiselBase(ModuleBase):
+
+    streams: int = 1
```

Comment:
Do we still need this one?
Isn't this covered by the `Port` attribute in any way?
If not what is the purpose of this attribute? I am a bit confused about this one.